src/compiler/scala/tools/nsc/backend/jvm/BCodeSkelBuilder.scala

```diff
@@ -96,26 +95,47 @@ abstract class BCodeSkelBuilder extends BCodeHelpers {
 
     /* ---------------- helper utils for generating classes and fields ---------------- */
 
-    def genPlainClass(cd: ClassDef): Unit = {
+    def genPlainClass(cd0: ClassDef): Unit = {
       assert(cnode == null, "GenBCode detected nested methods.")
 
-      claszSymbol       = cd.symbol
+      claszSymbol       = cd0.symbol
       isCZParcelable    = isAndroidParcelableClass(claszSymbol)
       isCZStaticModule  = isStaticModuleClass(claszSymbol)
       thisBType         = classBTypeFromSymbol(claszSymbol)
-      initModuleInClinit = isCZStaticModule && canAssignModuleInClinit(cd, claszSymbol)
 
       cnode = new ClassNode1()
 
       initJClass(cnode)
+      val cd = if (isCZStaticModule) {
+        // Move statements from the primary constructor following the superclass constructor call to
+        // a newly synthesised tree representing the "<clinit>", which also assigns the MODULE$ field.
+        // Because the assigments to both the module instance fields, and the fields of the module itself
+        // are in the <clinit>, these fields can be static + final.
 
-      val hasStaticCtor = methodSymbols(cd) exists (_.isStaticConstructor)
-      if (!hasStaticCtor) {
-        // but needs one ...
-        if (isCZStaticModule || isCZParcelable) {
-          fabricateStaticInit()
+        // TODO should we do this transformation earlier, say in Constructors? Or would that just cause
+        // pain for scala-{js, native}?
```

Comment:
@sjrd Thoughts?

Reply:
I haven't made my mind about it yet. I'll think about it.

Can thi my input wait until next week? I'm currently at the Scala Symposium.

Reply:
Sure thing.